### PR TITLE
faster conversion from Integer to BigInt

### PR DESCRIPTION
**Pull request by rfourquet**
This applies mainly to `BigInt(::[U]Int128)`, for which it's about 10 times faster. For isbits integers with 1024 bits, I observe a 20x improvement. The drawback is that this uses directly the `BigInt` internals (but we already do this sometimes, e.g. in `rand(BigInt)`).

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
GMP layout is documented and very stable so I don't think this is a problem. We also use it extensively in hashing of BigInts so many things would have to change if they were to change their layout.

**Comment by rfourquet** [comment]
I will merge in few days if no objection. The Appveyor failure is strange, many `StackOverflowError`s.

**Comment by simonbyrne** [comment]
I would be careful: I often see strange Win64 failures when changing MPFR or GMP code, but they end up being caused by real problems due to the `Clong` differences.

File: base/gmp.jl [hunk 1]
```diff
@@ -283,31 +283,21 @@ BigInt(x::Float16) = BigInt(Float64(x))
...
-        end
-        b = BigInt(0)
-        shift = 0
-        while x > 0
-            b += BigInt(UInt32(x&0xffffffff))<<shift
-            x >>>= 32
-            shift += 32
-        end
-        return b
+    x == 0 && return BigInt(0)
```
Comment (simonbyrne) [inline base/gmp.jl]: I think the problem is this line: it should be `BigInt(Clong(0))`.

**Comment by KristofferC** [comment]
> I will merge in few days if no objection. The Appveyor failure is strange, many StackOverflowErrors.

Unexplainable CI errors is an objection.

**Comment by rfourquet** [comment]
> Unexplainable CI errors is an objection.

Good point, I was a bit fast cataloguing this one as unrelated (the log was truncated on the web interface, it helped to download a texte version (which is still truncated apparently)).

**Comment by simonbyrne** [comment]
Thanks! Would be great to get the same for conversion to `BigFloat` as well.

**Comment by rfourquet** [comment]
> Would be great to get the same for conversion to BigFloat as well.

Good idea! But I don't think I will make a PR for this in the near future, as I'm not familiar with rounding modes. At least, the current implementation uses conversion to `BigInt` first, so will benefit from this PR already.

File: base/gmp.jl [hunk 2]
```diff
@@ -283,31 +283,21 @@ BigInt(x::Float16) = BigInt(Float64(x))
...
-            b += BigInt(UInt32(x&0xffffffff))<<shift
-            x >>>= 32
-            shift += 32
-        end
-        return b
+    x == 0 && return BigInt(Culong(0))
+    nd = ndigits(x, base=2)
+    z = MPZ.realloc2(nd)
+    s = sign(x)
+    s == -1 && (x = -x)
```
Comment (martinholters) [inline base/gmp.jl]: This fails for `x == typemin(x)`. Fix incoming.

Comment (simonbyrne) [reply base/gmp.jl]: really? I get the correct result (note that it uses unsigned shift `>>>` below).

Comment (simonbyrne) [reply base/gmp.jl]: to be fair, it might be cleaner to explicitly call `unsigned` here

Comment (simonbyrne) [reply base/gmp.jl]: and add `typemin`/`typemax` as test cases.

Comment (KristofferC) [reply base/gmp.jl]: https://github.com/JuliaLang/julia/pull/29340

Comment (rfourquet) [reply base/gmp.jl]: A small consolation for this embarassing bug is that it was [catched](https://github.com/JuliaLang/julia/pull/29238#issuecomment-423942246) by a recent test of mine :sweat_smile:

**Pull request merged**